core/services/ocr2/plugins/ocr2keeper/evm21/mercury/streams/streams.go

```diff
@@ -0,0 +1,330 @@
+package streams
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"math/big"
+	"net/http"
+	"sync"
+	"time"
+
+	"github.com/ethereum/go-ethereum/accounts/abi"
+	"github.com/ethereum/go-ethereum/accounts/abi/bind"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/common/hexutil"
+	"github.com/patrickmn/go-cache"
+
+	ocr2keepers "github.com/smartcontractkit/ocr2keepers/pkg/v3/types"
+
+	"github.com/smartcontractkit/chainlink-relay/pkg/services"
+
+	iregistry21 "github.com/smartcontractkit/chainlink/v2/core/gethwrappers/generated/i_keeper_registry_master_wrapper_2_1"
+	"github.com/smartcontractkit/chainlink/v2/core/logger"
+	"github.com/smartcontractkit/chainlink/v2/core/services/ocr2/plugins/ocr2keeper/evm21/core"
+	"github.com/smartcontractkit/chainlink/v2/core/services/ocr2/plugins/ocr2keeper/evm21/mercury"
+	v02 "github.com/smartcontractkit/chainlink/v2/core/services/ocr2/plugins/ocr2keeper/evm21/mercury/v02"
+	v03 "github.com/smartcontractkit/chainlink/v2/core/services/ocr2/plugins/ocr2keeper/evm21/mercury/v03"
+	"github.com/smartcontractkit/chainlink/v2/core/utils"
+)
+
+type Lookup interface {
+	Lookup(ctx context.Context, checkResults []ocr2keepers.CheckResult) []ocr2keepers.CheckResult
+}
+
+type latestBlockProvider interface {
+	LatestBlock() *ocr2keepers.BlockKey
+}
+
+type streamsRegistry interface {
+	GetUpkeepPrivilegeConfig(opts *bind.CallOpts, upkeepId *big.Int) ([]byte, error)
+	CheckCallback(opts *bind.CallOpts, id *big.Int, values [][]byte, extraData []byte) (iregistry21.CheckCallback, error)
+	Address() common.Address
+}
+
+type contextCaller interface {
+	CallContext(ctx context.Context, result interface{}, method string, args ...interface{}) error
+}
+
+type streams struct {
+	services.StateMachine
+	packer          mercury.Packer
+	mercuryConfig   mercury.MercuryConfigProvider
+	abi             abi.ABI
+	blockSubscriber latestBlockProvider
+	registry        streamsRegistry
+	client          contextCaller
+	lggr            logger.Logger
+	threadCtrl      utils.ThreadControl
+	v02Client       mercury.MercuryClient
+	v03Client       mercury.MercuryClient
+}
+
+// UpkeepPrivilegeConfig represents the administrative offchain config for each upkeep. It can be set by s_upkeepPrivilegeManager
+// role on the registry. Upkeeps allowed to use Mercury server will have this set to true.
+type UpkeepPrivilegeConfig struct {
+	MercuryEnabled bool `json:"mercuryEnabled"`
+}
+
+func NewStreamsLookup(
+	packer mercury.Packer,
+	mercuryConfig mercury.MercuryConfigProvider,
+	blockSubscriber latestBlockProvider,
```

Comment:
To get rid of the dependency of blockSubscriber, we created this `latestBlockProvider` interface, which requires latestBlock() func. This latestBlock func is the only func we need in streamsLookup.

Also refer to streams.go L#310